auth/role.go

```diff
@@ -0,0 +1,49 @@
+package auth
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Role represents the permission a user is given, if the Type is RoleSuperUser,
+// Then the user will have access to everything regardless of the value of Group.
+type Role struct {
+	Type  RoleType `json:"type"`
+	Group string   `json:"group"`
+}
+
+type RoleType string
+
+const (
+	RoleSuperUser = RoleType("super_user")
+	RoleUIAdmin   = RoleType("ui_admin")
+	RoleAdmin     = RoleType("admin")
+	RoleAPI       = RoleType("api")
+)
+
+func (r *RoleType) UnmarshalJSON(b []byte) error {
```

Comment:
So this a method that satisfies the json.Unmarshaler interface, when decoding JSON of any type, if the type implements this interface it'll be called instead of the json library attempting to decode the JSON automatically.